### PR TITLE
Add an API endpoint to profiling tools

### LINKED ISSUE #1305: Add an API endpoint to easily profile snapd
As a Snap contributor, I want to be able to easily use profiling tools, so I can generate clear views of the resources used by snapd, during a customized period.

### PR DESCRIPTION
**Pull request by kindermoumoute**
Fixes #1305, update documentation and add implementation of optional profiling tools endpoints in the Snap API.

Summary of changes:
- Add a flag to start profiling tools.
- When starting a plugin, pass the profiling tools flag.
- Add API endpoints for profiling tools in Snap and in the plugin side (new Go lib https://github.com/intelsdi-x/snap-plugin-lib-go/pull/44)

Testing done:
- start snapd with pprof flag `snapd -t 0 --pprof`.
- load plugins and start a task.
- curl snap API `curl "http://127.0.0.1:8181/v1/debug/pprof/profile?seconds=5" > cpu.pprof`
- `snapctl plugin list --running`, pprof port of running plugins should be displayed.
- curl plugin API on their pprof port `curl "http://127.0.0.1:<PORT>/v1/debug/pprof/profile?seconds=5" > cpu.pprof`

@intelsdi-x/snap-maintainers

### CODE REVIEW

File: docs/PROFILING.md [hunk 1]
```diff
@@ -59,56 +59,30 @@ go get github.com/uber/go-torch
...
-    signal.Notify(c, os.Interrupt, os.Kill, syscall.SIGTERM)
-
-    //Let's block until someone tells us to quit
-    go func() {
-        sig := <-c
-        p.Stop()    // Stop profiling
-        ...
+You can now get at any time a profile during a defined period (e.g. 120 seconds):
+```bash
+curl "http://127.0.0.1:8181/v1/debug/pprof/profile?seconds=120" > cpu.pprof
```
Comment (candysmurf) [inline docs/PROFILING.md]: please mention that whatever time specified, the curl command needs to wait for the time interval.

**Comment by kindermoumoute** [comment]
I'll add a demo gif in the documentation, to show what is supposed to happen.

File: mgmt/rest/server.go [hunk 2]
```diff
@@ -492,6 +500,15 @@ func (s *Server) addRoutes() {
 		s.r.GET("/v1/tribe/members", s.getMembers)
 		s.r.GET("/v1/tribe/member/:name", s.getMember)
 	}
+
+	// profiling tools routes
+	if s.pprof {
+		s.r.GET("/v1/debug/pprof/", s.index)
+		s.r.GET("/v1/debug/pprof/cmdline", s.cmdline)
+		s.r.GET("/v1/debug/pprof/profile", s.profile)
+		s.r.GET("/v1/debug/pprof/symbo", s.symbol)
```
Comment (candysmurf) [inline mgmt/rest/server.go]: can we use symbol in the GET URI as well.

**Comment by candysmurf** [comment]
I was able to run and see the torch.svg with Sierra.
screen shot 2016-10-31 at 3 20 17 pm


However, tasks were created and ran file with the following error:
```
▶ snapctl task create -t ~/task/task-mock-file.json
Using task manifest to create task
Error creating task:Unknown API response: invalid character 'P' looking for beginning of value

 Received: PANIC: runtime error: invalid memory address or nil pointer dereference
goroutine 19535 [running]:
github.com/intelsdi-x/snap/vendor/github.com/codegangsta/negroni.(*Recovery).ServeHTTP.func1(0xd13110, 0xc420396bc0, 0xc420130ac0)
	/Users/egu/.gvm/pkgsets/go1.7/snap-1.7/src/github.com/intelsdi-x/snap/vendor/github.com/codegangsta/negroni/recovery.go:34 +0xf2
panic(0x6d6f80, 0xc4200100b0)
	/Users/egu/.gvm/gos/go1.7/src/runtime/panic.go:458 +0x243
github.com/intelsdi-x/snap/mgmt/rest/rbody.assertSchedule(0xb15c80, 0xc42037e200, 0xc420224e60)
	/Users/egu/.gvm/pkgsets/go1.7/snap-1.7/src/github.com/intelsdi-x/snap/mgmt/rest/rbody/task.go:225 +0x2b6
github.com/intelsdi-x/snap/mgmt/rest/rbody.AddSchedulerTaskFromTask(0xb1d6e0, 0xc4201d28c0, 0xc42029d488)
	/Users/egu/.gvm/pkgsets/go1.7/snap-1.7/src/github.com/intelsdi-x/snap/mgmt/rest/rbody/task.go:111 +0x3d8
github.com/intelsdi-x/snap/mgmt/rest.(*Server).addTask(0xc4200df520, 0xd13110, 0xc420396bc0, 0xc4203d1c20, 0x0, 0x0, 0x0)
	/Users/egu/.gv
```

**Comment by kindermoumoute** [comment]
Documentation will come in another PR. https://github.com/intelsdi-x/snap-plugin-lib-go/pull/44 needs to be merged first so I can update glide.lock in this PR.

File: mgmt/rest/rbody/plugin.go [hunk 3]
```diff
@@ -107,4 +107,5 @@ type AvailablePlugin struct {
 	LastHitTimestamp int64  `json:"last_hit_timestamp"`
 	ID               uint32 `json:"id"`
 	Href             string `json:"href"`
+	PprofPort        string `json:"pprof_port"`
```
Comment (candysmurf) [inline mgmt/rest/rbody/plugin.go]: why is it using a string type for port?

File: control/available_plugin.go [hunk 4]
```diff
@@ -75,6 +75,7 @@ type availablePlugin struct {
 	exec               string
 	execPath           string
 	fromPackage        bool
+	pprofPort          string
```
Comment (candysmurf) [inline control/available_plugin.go]: string for port?

File: mgmt/rest/rbody/plugin.go (hunk 3 above)
Comment (kindermoumoute) [inline mgmt/rest/rbody/plugin.go]: At the end the goal is to display the port as a string when typing `snapctl plugin list --running`. Communication between snapctl and snapd are in json, which mean string anyway.

File: control/available_plugin.go (hunk 4 above)
Comment (kindermoumoute) [inline control/available_plugin.go]: Same reason, we have no reason to use it as an int. The goal is to displayed it at the end.

File: mgmt/rest/server.go [hunk 5]
```diff
@@ -60,6 +61,7 @@ const (
 	defaultAuth            bool   = false
 	defaultAuthPassword    string = ""
 	defaultPortSetByConfig bool   = false
+	defaultPprof           bool   = false
```
Comment (candysmurf) [inline mgmt/rest/server.go]: you don't need to set. The default is false. Of course, it does not hurt here.

File: mgmt/rest/server.go [hunk 6]
```diff
@@ -492,6 +507,41 @@ func (s *Server) addRoutes() {
 		s.r.GET("/v1/tribe/members", s.getMembers)
 		s.r.GET("/v1/tribe/member/:name", s.getMember)
 	}
+
+	// profiling tools routes
+	if s.pprof {
```
Comment (candysmurf) [inline mgmt/rest/server.go]: are you sure that you want to dynamically change it when snapd is already running? Will it need to restart snapd by doing this way?

File: mgmt/rest/server.go (hunk 5 above)
Comment (kindermoumoute) [inline mgmt/rest/server.go]: It doesn't change the binary (SSA optimize it), it's just to make it clear.

File: mgmt/rest/server.go (hunk 6 above)
Comment (kindermoumoute) [inline mgmt/rest/server.go]: > are you sure that you want to dynamically change it when snapd is already running? 

We don't enable/disable profiling tools without re-starting Snap. We do the same way than tribe. Else we would need another route to the API, in order to handle dynamic config changes. And I'm not sure we want that.

Comment (kindermoumoute) [inline mgmt/rest/server.go]: This feature is not about profiling tools, it's about all global option we can list with `snapd -h`. Currently we can't update those options without restarting Snap.

Comment (IRCody) [inline mgmt/rest/server.go]: I think adding the ability to dynamically update the config is outside the scope of this PR.

Comment (candysmurf) [inline mgmt/rest/server.go]: agreed that we can do dynamiclly update config in a different PR.
